Mathlib/RingTheory/Presentation.lean

```diff
@@ -0,0 +1,409 @@
+/-
+Copyright (c) 2024 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import Mathlib.RingTheory.Ideal.Cotangent
+
+/-!
+
+# Presentations of algebras
+
+## Main definition
+- `Algebra.Presentation`: A presentation of a `R`-algebra `S` consists of
+  1. `I`: The type of variables.
+  2. `val : I → S`: The assignment of each variable to a value.
+  3. `σ`: A section of `R[I] → S`.
+- `Algebra.Presentation.Hom`: Given a commuting square
+  ```
+  R --→ P = R[I] ---→ S
+  |                   |
+  ↓                   ↓
+  R' -→ P' = R'[I'] → S
+  ```
+  A hom between `P` and `P'` is an assignment `I → P'` such that the arrows commute.
+- `Algebra.Presentation.Cotangent`: The cotangent spance of a presentation `P = R[I] → S`, I.e. the
+  space `I/I²` with `I` being the kernel of the presentation.
+
+-/
+
+universe w u v
+
+open TensorProduct MvPolynomial
+
+variable (R : Type u) (S : Type v) [CommRing R] [CommRing S] [Algebra R S]
+
+/-- A presentation of a `R`-algebra `S` consists of
+1. `I`: The type of variables.
+2. `val : I → S`: The assignment of each variable to a value in `S`.
+3. `σ`: A section of `R[I] → S`. -/
+structure Algebra.Presentation where
```

Comment:
I am wondering about the naming scheme here: I would expect a finite presentation to be a `Presentation` in your sense where `I` is finite and the kernel of `R[I] → S` is finitely generated. So maybe only call a `Presentation` something where `I` is already finite? Then `Algebra.Presentation.Finite` could be the one with f.g. kernel.

Reply:
I need this API for not-necessarily finite algebras so I cannot require I to be finite.